ifrs17/Import/Importers.ipynb

```diff
@@ -509,6 +487,99 @@
+                "## Validate for Data Node States Logic"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public async static Task ValidateDataNodeStatesAsync(this IWorkspace workspace, Dictionary<string, DataNodeData> persistentDataNodeByDataNode)",
+                "\n{",
+                "\n    foreach(var importedDataNodeState in await workspace.Query<DataNodeState>().ToArrayAsync())",
+                "\n    {",
+                "\n        if(persistentDataNodeByDataNode.TryGetValue(importedDataNodeState.DataNode, out var currentPersistentDataNode))",
+                "\n        {",
+                "\n            if(importedDataNodeState.State < currentPersistentDataNode.State)",
+                "\n                ApplicationMessage.Log(Error.ChangeDataNodeState, importedDataNodeState.DataNode, ",
+                "\n                                                                  currentPersistentDataNode.State.ToString(), ",
+                "\n                                                                  importedDataNodeState.State.ToString());",
+                "\n",
+                "\n            if(importedDataNodeState.State == currentPersistentDataNode.State)",
+                "\n                await workspace.DeleteAsync<DataNodeState>(importedDataNodeState);",
+                "\n        }",
+                "\n    }",
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "## Validate for Mandatory Aoc Steps"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public async static Task ValidateForMandatoryAocSteps(this IWorkspace workspace, IDataSet dataSet, HashSet<AocStep> mandatoryAocSteps)",
+                "\n{",
+                "\n    var missingAocStepsByIdentityProperties = (await workspace.Query<RawVariable>().ToListAsync())",
+                "\n                     .GroupBy(x => IdentityPropertyReader<RawVariable>.Instance.GetIdentityString(x),",
+                "\n                              x => new AocStep(x.AocType, x.Novelty),",
+                "\n                              (properties, aocSteps) => (Properties: properties, AocSteps: mandatoryAocSteps.Except(aocSteps))",
+                "\n                             );",
+                "\n    foreach(var aocSteps in missingAocStepsByIdentityProperties) ",
+                "\n        foreach(var aocStep in aocSteps.AocSteps)  ApplicationMessage.Log(Warning.MandatoryAocStepMissing, aocStep.AocType, aocStep.Novelty, aocSteps.Properties);",
```

Comment:
foreach(var aocStep in aocSteps.AocSteps) --> foreach(var missingStep in missingSteps)
+
change the corresponding variables in the ApplicationMessage

Reply:
this is also outdated, resolving it